src/core/qgsmaprendererjob.cpp

```diff
@@ -354,6 +377,144 @@ LayerRenderJobs QgsMapRendererJob::prepareJobs( QPainter *painter, QgsLabelingEn
   return layerJobs;
 }
 
+LayerRenderJobs QgsMapRendererJob::prepareSecondPassJobs( LayerRenderJobs &firstPassJobs, LabelRenderJob &labelJob )
+{
+  LayerRenderJobs secondPassJobs;
+
+  static const QString labelLayerId = QStringLiteral( "**labels**" );
+
+  // We will need to quickly access the associated rendering job of a layer
+  QHash<QString, LayerRenderJob *> layerJobMapping;
+
+  // ... and whether a layer has a mask defined
+  QSet<QString> layerHasMask;
+
+  // We collect for each layer, the set of symbol layers that will be "masked"
+  // and the list of layers that have a mask
+  QHash<QString, QPair<QSet<QgsSymbolLayerId>, QList<QString>>> maskedSymbolLayers;
+  for ( LayerRenderJob &job : firstPassJobs )
+  {
+    QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( job.layer );
+    if ( ! vl )
+      continue;
+
+    layerJobMapping[job.layer->id()] = &job;
+
+    // collect label masks
+    auto labelMasks = QgsVectorLayerUtils::labelMasks( vl );
+    // collect other masks
+    auto symbolLayerMasks = QgsVectorLayerUtils::symbolLayerMasks( vl );
+
+    auto collectMasks = [&]( decltype( labelMasks ) * masks, QString sourceLayerId )
+    {
+      for ( auto it = masks->begin(); it != masks->end(); ++it )
+      {
+        auto lit = maskedSymbolLayers.find( it.key() );
+        if ( lit == maskedSymbolLayers.end() )
+        {
+          maskedSymbolLayers[it.key()] = qMakePair( it.value(), QList<QString>() << sourceLayerId );
+        }
+        else
+        {
+          if ( lit->first != it.value() )
+          {
+            QgsLogger::warning( QStringLiteral( "Layer %1 : Different sets of symbol layers are masked by different sources ! Only one (arbitrary) set will be retained !" ).arg( it.key() ) );
+            continue;
+          }
+          lit->second.push_back( sourceLayerId );
+        }
+      }
+      if ( ! masks->isEmpty() )
+        layerHasMask.insert( sourceLayerId );
+    };
+
+    collectMasks( &labelMasks, labelLayerId );
+    collectMasks( &symbolLayerMasks, vl->id() );
+  }
+
+  if ( maskedSymbolLayers.isEmpty() )
+    return secondPassJobs;
+
+  // Now that we know some layers have a mask, we have to allocate a mask image and painter
+  // for them in the first pass job
+  for ( LayerRenderJob &job : firstPassJobs )
+  {
+    QgsMapLayer *ml = job.layer;
+
+    if ( job.img == nullptr )
+    {
+      job.context.setPainter( allocateImageAndPainter( ml->id(), job.img ) );
```

Comment:
Also trying to understand the consequences of this -- does it mean that any layers set to be used as a mask themselves are forced to a raster output?